hydroflow_lang/src/graph/flat_graph_builder.rs

```diff
@@ -145,17 +220,103 @@ impl FlatGraphBuilder {
+                    Ok(code) => code,
+                    Err(err) => {
+                        self.diagnostics.push(err.clone().into());
+                        return syn::Result::Err(err);
+                    }
+                };
+
+                let flat_graph_builder = crate::graph::FlatGraphBuilder::from_hfmodule(statements);
+                let (flat_graph, _uses, diagnostics) = flat_graph_builder.build();
+                diagnostics
+                    .iter()
+                    .for_each(crate::diagnostic::Diagnostic::emit);
+
+                let mut ends = Ends {
+                    inn: None,
+                    out: None,
+                };
+
+                let mut node_mapping = BTreeMap::new();
+
+                for (nid, node) in flat_graph.nodes() {
+                    match node {
+                        Node::Operator(_) => {
+                            let varname = flat_graph.node_varname(nid);
+                            let new_id = self.flat_graph.insert_node(node.clone(), varname);
+                            node_mapping.insert(nid, new_id);
+                        }
+                        Node::ModuleBoundary { input, .. } => {
+                            let new_id = self.flat_graph.insert_node(
+                                Node::ModuleBoundary {
+                                    input: *input,
+                                    import_expr: import.span(),
+                                },
+                                Some(Ident::new(
+                                    &format!("module_{}", input.to_string()),
+                                    import.span(),
+                                )),
+                            );
+                            node_mapping.insert(nid, new_id);
+
+                            if *input {
+                                ends.inn = Some((
+                                    PortIndexValue::Elided(None),
+                                    GraphDet::Determined(new_id),
+                                ));
+                            } else {
+                                ends.out = Some((
+                                    PortIndexValue::Elided(None),
+                                    GraphDet::Determined(new_id),
+                                ));
+                            }
+                        }
+                        _ => panic!(),
+                    }
+                }
+
+                for (eid, (src, dst)) in flat_graph.edges() {
+                    let (src_port, dst_port) = flat_graph.edge_ports(eid);
+
+                    self.flat_graph.insert_edge(
+                        *node_mapping.get(&src).unwrap(),
+                        src_port.clone(),
+                        *node_mapping.get(&dst).unwrap(),
+                        dst_port.clone(),
+                    );
                 }
+
+                Ok(ends)
```

Comment:
Create a separate helper for this, something like `FlatGraphBuilder::merge_in(&mut self, other: FlatGraph)`